### PR TITLE
Patches and proposal for mainnet USDC market upgrade

### PR DESCRIPTION
**Pull request by jflatow**
This branch includes 2 recommended patches to the market logic, and creates a proposal to apply the updates to mainnet cUSDCv3.

*Emit erc20 transfer (mint) event when absorbed account ends up with a positive balance*
 - Calculate amount by converting from principal instead of directly using present value
 - Move Transfer event down
 - Fix unit test

*Automatically include excess balance in collateral reserves*

- Adds a reserves fn which it checks when collateral is bought

Note: we could add a fn to convert any (prior) collateral balance of the contract into reserves by clearing it out and removing from total. However, collateral reserves are currently 0, and we expect them to remain 0 if liquidation is working properly. If somehow collateral is acquired and held by the protocol (i.e. from an absorb) before this patch gets applied, the protocol will simply hold that balance as a supplier, until governance decides to move it, for which a fn can always be added (at that time) to assist.

Also note that we expect 1 scenario to fail on each of `mainnet`, `goerli`, and `fuji`, for one of two runs, due to testing for the newly patched in event. We also expect the new scenario testing that the entire ERC20 collateral balance is part of reserves fails in the run without the migration.

### CODE REVIEW

**Comment by jflatow** [comment]
TODO:
 - [ ] Scenario to transfer collateral to the market without supplying, and buy it back

**Comment by scott-silver** [comment]
Does this subsume #572 ?

**Comment by jflatow** [comment]
> Does this subsume #572 ?

Yes, was holding off to close that one until we merge this one, as its easier to review just the event change from that PR

File: deployments/fuji/usdc/migrations/1665028496_absorb_transfer_event.ts
Comment (scott-silver) [inline deployments/fuji/usdc/migrations/1665028496_absorb_transfer_event.ts:2]: nit: looks like `calldata`, `exp`, `expect` are all unused. same with the `COMP` on line 21

Comment (scott-silver) [reply deployments/fuji/usdc/migrations/1665028496_absorb_transfer_event.ts:2]: I guess we don't lint the `/deployments/` directory. I'll create a PR to enable that

File: deployments/fuji/usdc/migrations/1665028496_absorb_transfer_event.ts
Comment (scott-silver) [inline deployments/fuji/usdc/migrations/1665028496_absorb_transfer_event.ts:60]: is this a todo?

File: deployments/mainnet/usdc/migrations/1665028496_absorb_transfer_event.ts
Comment (scott-silver) [inline deployments/mainnet/usdc/migrations/1665028496_absorb_transfer_event.ts:61]: nit: everything excep `COMP` is unused

File: deployments/mainnet/usdc/migrations/1665028496_absorb_transfer_event.ts
Comment (scott-silver) [inline deployments/mainnet/usdc/migrations/1665028496_absorb_transfer_event.ts:42]: 🙌 that's awesome

File: scenario/LiquidationScenario.ts
Comment (scott-silver) [inline scenario/LiquidationScenario.ts:188]: looks like `ab0.events` can be null.

you could update this to: `ab0.events?.[2]?.event` if you wanted the expectations to fall when there were no events (instead of failing with a JS error about not finding `event` on `undefined`)

File: test/absorb-test.ts
Comment (scott-silver) [inline test/absorb-test.ts:360]: what are the 41K, 3K and 175 values here?

File: deployments/fuji/usdc/migrations/1665028496_absorb_transfer_event.ts
Comment (jflatow) [reply deployments/fuji/usdc/migrations/1665028496_absorb_transfer_event.ts:60]: No its there

File: scenario/LiquidationScenario.ts
Comment (jflatow) [reply scenario/LiquidationScenario.ts:188]: This is the test we want ;) See the note at the top

File: test/absorb-test.ts
Comment (jflatow) [reply test/absorb-test.ts:360]: These are from Kevin's branch, he updated this test to be more logical while he was at it - I believe those are coming from the prices?

File: deployments/fuji/usdc/migrations/1665028496_absorb_transfer_event.ts
Comment (scott-silver) [reply deployments/fuji/usdc/migrations/1665028496_absorb_transfer_event.ts:60]: is there a reason to call `getContracts()` then?

File: deployments/fuji/usdc/migrations/1665028496_absorb_transfer_event.ts
Comment (scott-silver) [reply deployments/fuji/usdc/migrations/1665028496_absorb_transfer_event.ts:2]: #601

File: scenario/LiquidationScenario.ts
Comment (jflatow) [reply scenario/LiquidationScenario.ts:188]: Good call 🦅

File: deployments/fuji/usdc/migrations/1665028496_absorb_transfer_event.ts
Comment (jflatow) [reply deployments/fuji/usdc/migrations/1665028496_absorb_transfer_event.ts]: I think I was going to check that the address had changed, but I removed this and updated the scen name / description

File: contracts/Comet.sol
Comment (kevincheng96) [inline contracts/Comet.sol:1209]: I think this change actually makes the protocol even more susceptible to a pre-transfer hook attack. A malicious token can bypass the `if (collateralAmount > getCollateralReserves(asset)) revert InsufficientReserves();` to buy more than the collateral reserves with re-entrancy.

This is because the accounting check now happens separately and before the `doTransferOut`, while the old code did the accounting check and `doTransferOut` together in `withdrawCollateral`.

Comment (jflatow) [reply contracts/Comet.sol:1209]: Malicious token can buy more of itself than reserves? Interesting, we should think this through

Comment (kevincheng96) [reply contracts/Comet.sol:1209]: Discussed offline. Thankfully, only the `msg.sender` (`Comet`) can register a pre-transfer hook for a ERC777 token. BUT, this is still vulnerable to a malicious token, so we should probably update the documentation.

Comment (jflatow) [reply contracts/Comet.sol:1209]: We think this isn't susceptible issues from ERC777 since only sender can register pre-transfer hooks in that standard. For some questionable asset in the future which allows arbitrary pre-transfer behavior, it should not be listed, but at least the risk is isolated to the questionable asset

File: contracts/Comet.sol
Comment (kevincheng96) [inline contracts/Comet.sol:480]: Worth documenting that we're intending for this to revert if `balanceOf(this) < totalSupplyAsset`?

Comment (jflatow) [reply contracts/Comet.sol:480]: Yep

File: scenario/LiquidationScenario.ts
Comment (kevincheng96) [inline scenario/LiquidationScenario.ts:188]: Cool, lets also check that this is a `mint` event as opposed to any random ERC20 transfer event:

```
expect(event(ab0, 2)).to.be.deep.equal({
      Transfer: {
        amount: baseBalance,
        from: ethers.constants.AddressZero,
        to: albert.address,
      }
    });
```

File: test/buy-collateral-test.ts
Comment (kevincheng96) [inline test/buy-collateral-test.ts:396]: Curious what these changes are for?

**Comment by kevincheng96** [comment]
Overall, this looks good. Only suggestions are to update the documentation in two places:
1. `getCollateralReserves` can revert under some cases, which is non-critical
2. A malicious token with a pre-transfer hook can buy users' collateral

File: test/buy-collateral-test.ts
Comment (jflatow) [reply test/buy-collateral-test.ts:396]: These are necessary bc you set up collateral reserves now differently than before. This is the amount of collateral that belongs to users excluding reserves now

Comment (kevincheng96) [reply test/buy-collateral-test.ts:396]: In that case, shouldn't these (`totalSupplyAsset`) stay at 100e18 while lines 421-430 change from `setCollateralBalance of Comet for WETH to 5e18` to `subtract 5e18 from totalSupplyAsset for WETH`?

Comment (jflatow) [reply test/buy-collateral-test.ts:396]: Hm, that's a good point, I guess the `setCollateralBalance` is supposed to be simulating absorb here, not a `supplyCollateral`... but I guess I don't really understand this part of the test. I can just delete the setting of totals collateral and collateral balance for the market, it doesn't change any of whats being tested, afaik

File: scenario/LiquidationScenario.ts
Comment (jflatow) [reply scenario/LiquidationScenario.ts:188]: This doesn't handle the pre-upgrade failure cases as gracefully, with an expectation failure like we made the other change to support, I think the presence of Transfer event should actually be sufficient, although could add a check for each of the properties if you really think its important?

**Review by kevincheng96 — APPROVED** [review]
👍

**Pull request merged**
